### PR TITLE
Btree errors serialize

### PR DESCRIPTION
**Pull request by RicoGit**

### CODE REVIEW

File: dataset/grpc/src/main/scala/fluence/dataset/grpc/DatasetStorageServer.scala [hunk 1]
```diff
@@ -98,42 +149,63 @@ class DatasetStorageServer[F[_] : Async](service: DatasetStorageRpc[F])(implicit
...
-      )
-    ).map(GetCallback(_))
+      valueF.attempt.flatMap {
+        case Right(value) ⇒
+          F.pure(GetCallback(GetCallback.Callback.Value(GetValue(value.fold(ByteString.EMPTY)(ByteString.copyFrom)))))
+        case Left(ClientError(msg)) ⇒
+          // when server recieve client reply with error, server shouln't answer anything
+          Async[F].raiseError[GetCallback](ClientError(msg))
+        case Left(exception) ⇒
+          F.pure(GetCallback(GetCallback.Callback.ServerError(Error(exception.getMessage))))
```
Comment (DieMyst) [inline dataset/grpc/src/main/scala/fluence/dataset/grpc/DatasetStorageServer.scala]: Maybe we should log all exceptions? With upper log level.

File: dataset/grpc/src/main/scala/fluence/dataset/grpc/DatasetStorageServer.scala [hunk 2]
```diff
@@ -98,42 +149,63 @@ class DatasetStorageServer[F[_] : Async](service: DatasetStorageRpc[F])(implicit
...
+    ): EitherT[Task, ClientError, T] = {
 
-    def getReply[T](check: PutCallbackReply.Reply ⇒ Boolean, extract: PutCallbackReply.Reply ⇒ T): Task[T] =
-      pull().map(_.reply).flatMap {
-        case r if check(r) ⇒ Task.now(extract(r))
-        case _             ⇒ Task.raiseError(new IllegalArgumentException("Wrong reply received, protocol error"))
-      }
+      val reply = clientReply()
+        .map {
+          case PutCallbackReply(reply) ⇒
```
Comment (DieMyst) [inline dataset/grpc/src/main/scala/fluence/dataset/grpc/DatasetStorageServer.scala]: Here suspicious shadowing shows by code analizer

File: dataset/grpc/src/main/scala/fluence/dataset/grpc/DatasetStorageServer.scala [hunk 3]
```diff
@@ -34,30 +37,77 @@ import monix.reactive.Observer
...
-  private def runT[A](fa: Task[A]): F[A] = fa.toIO.to[F]
+  /**
+   * Convert function: {{{ EitherT[Task, E, V] => F[V] }}}.
+   * It's temporary decision, it will be removed when EitherT[F, E, V] was everywhere.
+   *
+   * @tparam F Type of effect with [[cats.effect.Async]] bound
+   * @tparam E Type of exception, should be subclass of [[Throwable]]
+   * @tparam V Type of value
+   */
+  private def runT[F[_] : Async, E <: Throwable, V](eitherT: EitherT[Task, E, V]): F[V] = eitherT.value.flatMap {
```
Comment (DieMyst) [inline dataset/grpc/src/main/scala/fluence/dataset/grpc/DatasetStorageServer.scala]: F is already in class constructor

File: dataset/grpc/src/main/scala/fluence/dataset/grpc/DatasetStorageServer.scala (hunk 1 above)
Comment (RicoGit) [reply dataset/grpc/src/main/scala/fluence/dataset/grpc/DatasetStorageServer.scala]: yes

File: dataset/grpc/src/main/scala/fluence/dataset/grpc/DatasetStorageServer.scala (hunk 2 above)
Comment (RicoGit) [reply dataset/grpc/src/main/scala/fluence/dataset/grpc/DatasetStorageServer.scala]: thx

File: dataset/grpc/src/main/scala/fluence/dataset/grpc/DatasetStorageServer.scala (hunk 3 above)
Comment (RicoGit) [reply dataset/grpc/src/main/scala/fluence/dataset/grpc/DatasetStorageServer.scala]: thx

File: dataset/grpc/src/main/scala/fluence/dataset/grpc/DatasetStorageClient.scala [hunk 4]
```diff
@@ -52,52 +64,86 @@ class DatasetStorageClient[F[_] : Effect](
...
-
-        getCallbacks
-          .submitLeaf(sl.keys.map(_.toByteArray).toArray, sl.valuesChecksums.map(_.toByteArray).toArray)
-
-          .map(oi ⇒ ReplySubmitLeaf(oi.getOrElse(-1)))
-          .map(GetCallbackReply.Reply.SubmitLeaf)
-
-    }.map(GetCallbackReply(_))
+    /** Puts error to client error(for returning error to user of this client), and return reply with error for server.*/
+    def handleClientErr(err: Throwable): F[GetCallbackReply] = {
```
Comment (alari) [inline dataset/grpc/src/main/scala/fluence/dataset/grpc/DatasetStorageClient.scala]: For the case of client error, shouldn't the `pipe` be closed? I don't see any code for doing it

File: dataset/grpc/src/main/scala/fluence/dataset/grpc/DatasetStorageClient.scala [hunk 5]
```diff
@@ -52,52 +64,86 @@ class DatasetStorageClient[F[_] : Effect](
...
+
+      }
 
     (
       Observable(
-        GetCallbackReply(GetCallbackReply.Reply.DatasetId(
-          DatasetId(ByteString.copyFrom(datasetId))
-        ))
+        GetCallbackReply(
+          GetCallbackReply.Reply.DatasetId(DatasetId(ByteString.copyFrom(datasetId)))
```
Comment (alari) [inline dataset/grpc/src/main/scala/fluence/dataset/grpc/DatasetStorageClient.scala]: Not sure if this formatting is more readable

File: dataset/grpc/src/main/scala/fluence/dataset/grpc/DatasetStorageServer.scala [hunk 6]
```diff
@@ -66,7 +116,7 @@ class DatasetStorageServer[F[_] : Async](service: DatasetStorageRpc[F])(implicit
            * @param valuesChecksums Checksums of values for current leaf
            * @return index of searched value, or None if key wasn't found
            */
-          override def submitLeaf(keys: Array[Key], valuesChecksums: Array[Hash]): F[Option[Int]] =
+          override def submitLeaf(keys: Array[Key], valuesChecksums: Array[Hash]): F[Option[Int]] = {
```
Comment (alari) [inline dataset/grpc/src/main/scala/fluence/dataset/grpc/DatasetStorageServer.scala]: Are you sure it should be committed?

File: dataset/grpc/src/main/scala/fluence/dataset/grpc/DatasetStorageServer.scala [hunk 7]
```diff
@@ -98,42 +149,66 @@ class DatasetStorageServer[F[_] : Async](service: DatasetStorageRpc[F])(implicit
...
+      }
+    )
 
     stream
   }
 
   override def put(responseObserver: StreamObserver[PutCallback]): StreamObserver[PutCallbackReply] = {
     val resp: Observer[PutCallback] = responseObserver
     val (repl, stream) = streamObservable[PutCallbackReply]
+    val clientReply = repl.pullable
```
Comment (alari) [inline dataset/grpc/src/main/scala/fluence/dataset/grpc/DatasetStorageServer.scala]: The meaning of an action is lost with this renaming. `clientReply` is a noun, while `pull` is a verb. Pull means that you request a pulling action and wait while it completes. Without asking to pull, the pull doesn't happen, and hence backpressure appears. I'd suggest to name it `pullReply` or `pullClientReply` or smth like that.

File: dataset/grpc/src/main/scala/fluence/dataset/grpc/DatasetStorageClient.scala (hunk 4 above)
Comment (RicoGit) [reply dataset/grpc/src/main/scala/fluence/dataset/grpc/DatasetStorageClient.scala]: How I can do that?

Comment (alari) [reply dataset/grpc/src/main/scala/fluence/dataset/grpc/DatasetStorageClient.scala]: After pushing the client error into response stream, close the stream

File: dataset/grpc/src/main/scala/fluence/dataset/grpc/DatasetStorageServer.scala (hunk 7 above)
Comment (RicoGit) [reply dataset/grpc/src/main/scala/fluence/dataset/grpc/DatasetStorageServer.scala]: Ok

File: dataset/grpc/src/main/scala/fluence/dataset/grpc/DatasetStorageClient.scala (hunk 4 above)
Comment (RicoGit) [reply dataset/grpc/src/main/scala/fluence/dataset/grpc/DatasetStorageClient.scala]: Will do it in next PR

File: dataset/grpc/src/main/scala/fluence/dataset/grpc/DatasetStorageServer.scala (hunk 7 above)
Comment (RicoGit) [reply dataset/grpc/src/main/scala/fluence/dataset/grpc/DatasetStorageServer.scala]: ok

**Pull request merged**